README.md
# rpki-rtr-client
## INTRODUCTION
A simple client-side implementation of the RTR [RFC8210](https://tools.ietf.org/html/rfc8210) protocol in Python.
See Cloudflare's blogs on [RPKI](https://blog.cloudflare.com/tag/rpki/) for more open source RPKI code.
## INSTALL
Presently the code is easiest to install via github.
```
   $ pip3 install pytricia
   ...
   $

   $ git clone https://github.com/cloudflare/rpki-rtr-client.git
   ...
   $ cd rpki-rtr-client
   $
```
The [pytricia](https://pypi.org/project/pytricia/) package is used for storing a routing table.
## USAGE
Cloudflare provides an open RTR server at `rtr.rpki.cloudflare.com` port `8282` based on Cloudflare's [GoRTR](https://github.com/cloudflare/gortr) open source RTR server.

Usage is via the `--help` argument.
```
   $ rtr_client/rtr_client.py --help
   usage: rtr_client [-H|--help] [-v|--verbose] [-h|--host] hostname [-p|--port] portnumber [-s|--serial] serialnumber [-t|--timeout] seconds [-d|--dump] 
   $
```

The Cloudflare open RTR server default hostname and port are compiled into the source code.
You can specify your own host and port via the command line `-h|--host` and `-p|--port` arguments.
```
   $ rtr_client/rtr_client.py --host rtr.rpki.cloudflare.com --port 8282
   ...
   ^C
   $
```
A modocom of debug information is available to show the serial number and the progress of accepting announce/widthdraw valid ROAs.
The code will always show the progress of serial numbers plus the number of valid ROAs associated with that serial number.
```
   DUMP ROUTES: serial=133 announce=130401/withdraw=0
   NEW SERIAL 0->133
```
As the protocol continues to sync, the messages show progress on the routing list.
```
   DUMP ROUTES: serial=381 announce=18/withdraw=2
   NEW SERIAL 380->381
```
The `.` debug message simply mean that PDUs have been transfered between RTR server and RTR client.

## DATA FILES
There's a data directory created with JSON files of every serial numbers worth of ROA data.
The directory is sorted by `YYYY-MM` and the files include the full date (in UTC).
```
   $ ls -lt data/
   total 0
   drwxr-xr-x  7 martin martin  224 Feb 11 09:36 2020-02
   $

   $ ls -lt data/2020-02
   total 21592
   -rw-r--r--  1 martin martin  5520676 Feb 16 18:22 2020-02-17-022209.routes.00000365.json
   -rw-r--r--  1 martin martin  5520676 Feb 16 18:42 2020-02-17-024242.routes.00000838.json
   -rw-r--r--  1 martin martin      412 Feb 16 19:56 2020-02-17-035645.routes.00000841.json
   -rw-r--r--  1 martin martin      272 Feb 16 20:16 2020-02-17-041647.routes.00000842.json
   -rw-r--r--  1 martin martin      643 Feb 16 20:36 2020-02-17-043649.routes.00000843.json
   $
```
You can review those files for how many RTR announce/withdraw ROAs were processed.
```
   $ for f in data/2020-02/*.json ; do echo "$f `jq -r '.routes.announce[]|.ip' < $f | wc -l` `jq -r '.routes.withdraw[]|.ip' < $f | wc -l`" ; done
   data/2020-02/2020-02-17-022209.routes.00000365.json   128483        0
   data/2020-02/2020-02-17-024242.routes.00000838.json   128483        0
   data/2020-02/2020-02-17-035645.routes.00000841.json        3        6
   data/2020-02/2020-02-17-041647.routes.00000842.json        5        0
   data/2020-02/2020-02-17-043649.routes.00000843.json        9        5
   $
```
You can list the ROAs. A `null` mean that no MaxLen has been specified in the ROA.
```
   $ jq -r '.routes.announce[]|.ip,.asn,.maxlen' data/2020-02/*0838.json | paste - - - | sort -V | head
   1.0.0.0/24      13335   null
   1.1.1.0/24      13335   null
   1.9.0.0/16      4788    24
   1.9.12.0/24     65037   null
   1.9.21.0/24     24514   null
   1.9.23.0/24     65120   null
   1.9.31.0/24     65077   null
   1.9.65.0/24     24514   null
   1.34.0.0/15     3462    24
   1.36.0.0/16     4760    null
   $
```
Additionally, the full list of valid ROAs is dumped into `data/routingtable.json` which can then be used the `show` command:
```
   $ rtr_client/rtr_show.py 1.37.0.0/16 112.198.0.0/16
   ROUTE            ROA              MaxLen ASN
   1.37.0.0/16      1.37.0.0/16         /17 AS4775
   ROUTE            ROA              MaxLen ASN
   112.198.0.0/16   112.198.0.0/16      /24 AS4775
   $
```
The `-l` argument will show add more specific ROAs.

The code can also dump the raw binary protocol and then replay that data to debug the protocol with the `-d|--dump` argument.
This generates a `data/__________-raw-data.bin` file.
The `file_process.py` command will process that file.

## CHANGELOG
 - This is the first release and while it works, it is not ready for prime time
 - Directory format updated to split by YYYY-MM in case it gets big (plus the serial number may not be sequential)
 - Moved from 3rd party `netaddr` package to Python's `ipaddress` data type
 - All internal cidr's are stored as `ipaddress` types and JSON processing now handles that type correctly
 - Added valid route table and show command
 - Renamed show.py to rtr_show.py
 - moved code to rtr_client folder
 - Added -V/--version flags
 - Added support for tracking session_id's
 - Fixed route dump duplication after session restart
 - First pass at pypi packaging
 - Cleaned up route processing
 - Timestamp added to major debug messages
 - Connect class does all the socket processing now - just cleaner that way

## LICENSE
Licensed under the BSD 3 License. See [LICENSE.txt](LICENSE.txt) file.



CHANGELOG.md
# Change Log

 - 2020-06-06 14:52:54 -0700 [1c065a3](../../commit/1c065a3faa285ada7f33445a7173d9373c46d240) CHANGELOG updated
 - 2020-06-06 14:51:59 -0700 [bb8ad13](../../commit/bb8ad13e51881ad7f81d9f135940e9b781f72446) added CHANGELOG file
 - 2020-06-06 14:50:24 -0700 [de0a49b](../../commit/de0a49bcb8a46a1b9b82dc5580ee54012b41c5c3) 1.2.0 release
 - 2020-06-06 14:46:48 -0700 [a34dc27](../../commit/a34dc2711d5d1f0c9c9682152f04fbf376939c27) pylint, session id, router key, binary protocol code, debug strings  - all improved
 - 2020-06-06 14:44:49 -0700 [f1bdc8f](../../commit/f1bdc8f5a90b1adbc77aa9cbe69faab34774b9a1) pylint, session id, socket connect, ^C handling - all improved
 - 2020-06-06 12:12:48 -0700 [0e4c579](../../commit/0e4c5798b13c468c60d58811a8890b555d078c78) pylint kept happy - well somewhat happy
 - 2020-06-06 12:12:36 -0700 [c50ad98](../../commit/c50ad98a2e361e5e858374006422dabaa8d35001) pylint kept happy - well somewhat happy
 - 2020-06-06 12:12:11 -0700 [5a16aef](../../commit/5a16aeffb4e0aec28debe1b6b7ae40eaafaee48c) pylint kept happy - well somewhat happy
 - 2020-06-06 12:11:44 -0700 [333c0bb](../../commit/333c0bb67bdb3a7e2aae6285aec65d4792d07a6d) pylint kept happy - well somewhat happy
 - 2020-06-06 12:11:27 -0700 [10e0ae3](../../commit/10e0ae321621073a5e5a608cdb235074eaa0d9e3) pylint kept happy - well somewhat happy
 - 2020-06-06 12:11:04 -0700 [911d6dc](../../commit/911d6dc445793b2be2398cb1a77c01f4e2857e6c) pylint kept happy - well somewhat happy
 - 2020-06-06 11:53:41 -0700 [393a0f2](../../commit/393a0f2f214ab2eb2364d3ab2fb4e3c721dcdbe8) pylint kept happy - well somewhat happy
 - 2020-06-05 17:45:28 -0700 [b0096cd](../../commit/b0096cdce6adb617f7d3b59dd978cb8232dec53d) added debug for tracking down SESSION ID issues
 - 2020-06-04 12:22:05 -0700 [f2dc421](../../commit/f2dc421cc121079952f4237a78bbd7d7989aa920) removed unused README build code
 - 2020-06-04 12:19:47 -0700 [e6f0b81](../../commit/e6f0b8179c25b78f64ad0369636d890eed0572d0) minor update to usage string
 - 2020-06-04 12:15:34 -0700 [1041e84](../../commit/1041e84165d6edb46c6d93cfd005873da6cad500) Merge pull request #7 from oskar456/fix_pypi_package
 - 2020-06-04 12:09:31 -0700 [34989a2](../../commit/34989a22208d9f2689a3ca0dea692b28af70f833) upped version number after pull/8
 - 2020-06-04 12:06:56 -0700 [69a9bf9](../../commit/69a9bf9372e74c3aabfc63ee716cc4e854a01cba) Merge pull request #8 from gmarsay/ipv6-support
 - 2020-06-04 12:05:36 -0700 [158000d](../../commit/158000d84e9c4ae3f6ed880565d3310d850f8442) added and fixed support for Session ID - see RFC
 - 2020-06-04 17:35:03 +0200 [22058a5](../../commit/22058a560d0f6b603e17664b5318ad396f982cc0) Add support for IPv6
 - 2020-06-02 09:56:17 +0200 [1fef4ae](../../commit/1fef4aee3595426debd3f91951c5e3806439fccf) Keep only RST documentation, fix command line help
 - 2020-06-02 09:31:35 +0200 [a5c7eeb](../../commit/a5c7eeb87ddeebe22ed0586c9e29c0d67224851c) Make the package work out of the box when installed with pip3
 - 2020-05-12 17:07:11 -0700 [7a8c723](../../commit/7a8c7234b5e13b26513cb46b4c59306b26d8653c) 1.0.5 release - still testing
 - 2020-05-12 17:05:08 -0700 [014dbca](../../commit/014dbca6cf5a0dbdf0e94ff4d30cb1350572bb15) cleaned up use of ipaddress types - actually less code
 - 2020-04-22 14:12:05 -0700 [0e86a4c](../../commit/0e86a4c911e9d28a257b33762fde734ad08ef9fe) 1.0.4 release - still testing
 - 2020-04-22 14:10:04 -0700 [f8f21c1](../../commit/f8f21c11f7cb73ecda0e473c9305648f8509b9f4) cleaned up route processing, timestamps in message
 - 2020-04-06 11:36:31 -0700 [5c081ca](../../commit/5c081ca8f5935b69c5265e5f9fe76d112aaa36d4) moved code to rtr_client folder, added session_id code, renamed show to rtr_show, fixed dump duplication
 - 2020-03-04 20:43:12 -0800 [ced1ec6](../../commit/ced1ec6aa29931d93ecbfc2fcc2d2c993ab08e68) updated and corrected license - issue #2
 - 2020-02-28 23:47:21 -0800 [c814ac7](../../commit/c814ac7a198393ef0b3a2f7f73253006a81c80e4) pypi packaging added - but not complete
 - 2020-02-28 23:46:23 -0800 [bda6258](../../commit/bda62580f177e99d6e2c6bfcec72fdcdec5efa35) pypi packaging added
 - 2020-02-28 23:12:35 -0800 [9f92c32](../../commit/9f92c32d46c388883558a63e895e5b9d9e53a9e2) /data/ correctly defined
 - 2020-02-28 23:11:21 -0800 [b52616a](../../commit/b52616a092d70701ca1ac4bebeed6f10dfca2204) added more documentation
 - 2020-02-28 23:11:07 -0800 [21ac8b3](../../commit/21ac8b353b2dca6805735aa2f1a35a276c922a5d) typo
 - 2020-02-28 23:04:51 -0800 [9422f23](../../commit/9422f23857b97e7fe863eb2a02b1dae4c65fe2c7) added ROA table and show command. moved from netaddr to ipaddress. json file now pretty. added more documentation.
 - 2020-02-19 20:36:55 -0800 [d17308e](../../commit/d17308e0e2c32163c7d68c0e52bcc582e6bb9a18) decided to cleanly format json file. added new route dump logic
 - 2020-02-17 13:48:31 -0800 [9eb9122](../../commit/9eb9122bd2ba057c60fe4d96af00324d4225d2f9) Merge pull request #1 from ties/fix/data_directory_requires_name
 - 2020-02-17 14:43:11 +0100 [562a884](../../commit/562a884efff0691e7b86ead20883b5d7d4039d51) Pass in now to data_directory function
 - 2020-02-16 21:21:51 -0800 [15726eb](../../commit/15726ebaf5e293e212794de1f7487ac337ea1bf8) new data directory format
 - 2020-02-16 21:15:41 -0800 [572d4f5](../../commit/572d4f5fbca11f1a704ce34782f9b89a691a8a77) new data directory format
 - 2020-02-16 21:14:13 -0800 [1ba11e0](../../commit/1ba11e096b698f9a8b63be48a5995624cb75cf58) new data directory format
 - 2020-02-16 21:13:29 -0800 [54cd92f](../../commit/54cd92fece84b05d52044ed4cea5c3f8f8b8f9cd) misc additions. new data directory format
 - 2020-02-16 15:32:17 -0800 [d66b172](../../commit/d66b1728e134db39f556f4102db9aa743a0badc2) typo when writing data for non-maxlen units
 - 2020-02-11 18:51:49 -0800 [dda6c7b](../../commit/dda6c7b17e6b825fd55adaae7afbc80cc2071822) ignore
 - 2020-02-11 18:48:36 -0800 [f7301b6](../../commit/f7301b6f4b1488d31083475f273f789d7dbc0c03) fixed debug param - its int vs bool

rtr_client/rtr_client.py
#!/usr/bin/env python3
"""RTR client"""

import sys
import os
import getopt
import socket
import select
import time
import json
import ipaddress
from datetime import datetime
from random import randrange

try:
	import pytricia
except:
	pytricia = None

try:
	from rtr_protocol import rfc8210router
	from __init__ import __version__
except ImportError:
	from .rtr_protocol import rfc8210router
	from .__init__ import __version__

#
# rtr protocol - port 8282 - clear text - Cisco, Juniper
# rtr_protocol - port 8283 - ssh - Juniper
# rtr_protocol - port 8284 - tls - ?
#

class Connect(object):
	"""RTR client"""

	rtr_host = 'rtr.rpki.cloudflare.com'
	rtr_port = 8282
	fd = None
	connect_timeout = 5 # this is about the socket connect timeout and not data timeout

	def __init__(self, host=None, port=None):
		"""RTR client"""
		if host:
			self.rtr_host = host
		if port:
			self.rtr_port = port
		self.fd = self._connect()

	def close(self):
		"""RTR client"""
		self.fd.close()
		self.fd = None

	def recv(self, n):
		"""RTR client"""
		try:
			return self.fd.recv(n)
		except Exception as e:
			raise

	def send(self, packet):
		"""RTR client"""
		try:
			return self.fd.send(packet)
		except BrokenPipeError as e:
			# remote end closed connection
			sys.stderr.write('send: Broken Pipe %s\n ' % (e))
			sys.stderr.flush()
			raise
		except Exception as e:
			sys.stderr.write('send: Error %s\n ' % (e))
			sys.stderr.flush()
			raise

	def name(self):
		"""RTR client"""

		if self._sockaddr:
			return '%s.%s' % socket.getnameinfo(self._sockaddr, 0)
		raise ValueError

	def _sleep(self, n):
		"""RTR client"""
		# simple back off for failed connect
		time.sleep(n)

	def _connect(self):
		"""RTR client"""
		try:
			ginfo = socket.getaddrinfo(self.rtr_host, self.rtr_port, 0, 0, socket.SOL_TCP)
		except socket.gaierror as e:
			sys.stderr.write('socket: %s.%s: %s (%d)\n' % (self.rtr_host, self.rtr_port, str(e.strerror), int(e.errno)))
			sys.stderr.flush()
			sys.exit(1)

		for ii in [1, 1, 2, 4, 8, 16, 32]:
			for gthis in ginfo:
				try:
					afamily, socktype, proto, canonname, sockaddr = gthis
					fd = socket.socket(afamily, socktype, proto)
					fd.settimeout(self.connect_timeout)
					fd.connect(sockaddr)
					self._sockaddr = sockaddr
					self.fd = fd
					return fd
				except socket.timeout:
					sys.stderr.write('socket: %s.%s: connection timeout\n' % (sockaddr[0], sockaddr[1]))
					sys.stderr.flush()
					self._sleep(ii)
					continue
				except socket.error as e:
					sys.stderr.write('socket: %s.%s: %s (%d)\n' % (sockaddr[0], sockaddr[1], str(e.strerror), int(e.errno)))
					sys.stderr.flush()
					self._sleep(ii)
					continue

		self._sockaddr = None
		self.fd = None
		return None

class Process(object):
	"""RTR client"""

	class Buffer(object):
		"""RTR client"""

		def __init__(self):
			"""RTR client"""
			self.last_buffer = None

		def clear(self):
			"""RTR client"""
			if self.last_buffer:
				self.last_buffer = None
		def read(self):
			"""RTR client"""
			b = self.last_buffer
			if self.last_buffer:
				self.last_buffer = None
			return b

		def write(self, b):
			"""RTR client"""
			self.last_buffer = b

	def __init__(self):
		"""RTR client"""
		self.buf = self.Buffer()

	def do_hunk(self, rtr_session, v):
		"""RTR client"""
		if not v or len(v) == 0:
			# END OF FILE
			return False

		b = self.buf.read()
		if b:
			v = b + v
		data_length = len(v)
		data_left = rtr_session.process(v)
		if data_left > 0:
			self.buf.write(v[data_length - data_left:])
		return True

	def clear(self):
		"""RTR client"""
		self.buf.clear()

def now_in_utc():
	"""RTR client"""

	now = datetime.utcnow().replace(tzinfo=None)
	# YYYY-MM-DD-HHMMSS
	return now.strftime('%Y-%m-%d-%H%M%S')

def data_directory(now):
	"""RTR client"""

	# data/YYYY-MM
	try:
		os.mkdir('data')
	except FileExistsError:
		pass
	try:
		os.mkdir('data' + '/' + now[0:7])
	except FileExistsError:
		pass

def dump_routes(rtr_session, serial, session_id):
	"""RTR client"""

	# dump present routes into file based on serial number and session_id
	routes = rtr_session.routes()
	if len(routes['announce']) > 0 or len(routes['withdraw']) > 0:
		now = now_in_utc()
		data_directory(now)
		j = {'serial': serial, 'session_id': session_id, 'routes': routes}
		with open('data/%s/%s.routes.%08d.%08d.json' % (now[0:7], now, session_id, serial), 'w') as fd:

			class IPAddressEncoder(json.JSONEncoder):
				"""RTR client"""

				def default(self, obj):
					"""RTR client"""
					if pytricia and isinstance(obj, pytricia.PyTricia):
						a = {}
						for prefix in obj:
							a[prefix] = obj[prefix]
						return a
					if isinstance(obj, ipaddress.IPv4Network):
						return str(obj)
					if isinstance(obj, ipaddress.IPv6Network):
						return str(obj)
					return json.JSONEncoder.default(self, obj)

			fd.write(json.dumps(j, indent=2, cls=IPAddressEncoder))

		# clean up from this serial number
		rtr_session.clear_routes()
		sys.stderr.write('%s: DUMP ROUTES: session_id=%d serial=%d announce=%d/withdraw=%d\n' % (
						now_in_utc(), session_id, serial, len(routes['announce']), len(routes['withdraw'])))
		sys.stderr.flush()

		# dump the full routing table
		rtr_session.save_routing_table()

def rtr_client(host=None, port=None, serial=None, session_id=None, timeout=None, dump=False, debug=0):
	"""RTR client"""

	rtr_session = rfc8210router(serial=serial, session_id=session_id, debug=debug)

	if dump:
		data_directory(now_in_utc())
		dump_fd = open('data/__________-raw-data.bin', 'w')

	p = Process()

	have_session_id = False

	connection = None
	while True:
		if not connection:
			try:
				connection = Connect(host, port)
			except KeyboardInterrupt:
				# no need to print anything - just exit!
				sys.exit(1)

		if not connection.fd:
			connection = None
			sys.stderr.write('%s: NO NETWORK CONNECTION\n' % (now_in_utc()))
			sys.stderr.flush()
			# sys.exit(1)
			continue

		p.clear()
		have_session_id = False
		sys.stderr.write('%s: CONNECT %s\n' % (now_in_utc(), connection.name()))
		sys.stderr.flush()

		if session_id is None or session_id == 0 or serial is None or serial == 0:
			# starting from scratch!
			packet = rtr_session.reset_query()
			serial = 0
			have_session_id = False
			session_id = 0
		else:
			# packet = rtr_session.serial_query(serial)
			packet = rtr_session.serial_query()

		# send the first packet on the connection -- kicking things off!
		try:
			sys.stderr.write('+')
			sys.stderr.flush()
			connection.send(packet)
		except Exception as e:
			sys.stderr.write('send: %s\n' % (e))
			sys.stderr.flush()
			connection.close()
			connection = None
			# this will open up a fresh connection and try all over again
			continue

		while True:
			# At every oppertunity, see if we have a new session_id number
			try:
				new_session_id = rtr_session.get_session_id()
				if have_session_id:
					if new_session_id != session_id:
						sys.stderr.write('\n%s: REFRESHED SESSION ID %d->%d\n' % (now_in_utc(), session_id, new_session_id))
						sys.stderr.flush()
						# consider a reset here - once we handle 0
				else:
					sys.stderr.write('\n%s: NEW SESSION ID %d\n' % (now_in_utc(), new_session_id))
					sys.stderr.flush()
				# update session_id number
				session_id = new_session_id
				have_session_id = True
			except ValueError:
				# no session_id number known yet - should only happen once
				sys.stderr.write('%s: NO SESSION ID\n' % (now_in_utc()))
				sys.stderr.flush()
				pass

			# At every oppertunity, see if we have a new serial number
			new_serial = rtr_session.cache_serial_number()
			if new_serial != serial:
				try:
					new_session_id = rtr_session.get_session_id()
				except ValueError:
					new_session_id = 0
				sys.stderr.write('\n%s: SESSION %d NEW SERIAL %s->%d\n' % (now_in_utc(), new_session_id, serial, new_serial))
				sys.stderr.flush()
				# dump present routes into file based on serial number
				dump_routes(rtr_session, new_serial, new_session_id)
				# update serial number
				serial = new_serial
				# update session_id
				session_id = new_session_id

			try:
				# because random timers are your friend! but keep above one second - just because
				delta = 0.2
				this_timeout = max(1.0, float(randrange(timeout * (1-delta), timeout * (1+delta), 1)))
				ready = select.select([connection.fd], [], [], this_timeout)
			except KeyboardInterrupt:
				sys.stderr.write('\nselect wait: ^C\n')
				sys.stderr.flush()
				sys.exit(1)
			except Exception as e:
				sys.stderr.write('\nselect wait: %s\n' % (e))
				sys.stderr.flush()
				break

			if not ready[0]:
				# Timeout
				sys.stderr.write('T')
				sys.stderr.flush()

				if rtr_session.time_remaining():
					sys.stderr.write('-')
					sys.stderr.flush()
					continue

				# timed out - go ask for more data!
				packet = rtr_session.serial_query()
				## rtr_session.process(packet)
				try:
					sys.stderr.write('s')
					sys.stderr.flush()
					connection.send(packet)
					continue
				except Exception as e:
					sys.stderr.write('send: %s\n' % (e))
					sys.stderr.flush()
					connection.close()
					connection = None
					break

			try:
				sys.stderr.write('.')
				sys.stderr.flush()
				v = connection.recv(64*1024)
			except Exception as e:
				sys.stderr.write('recv: %s\n' % (e))
				sys.stderr.flush()
				v = None
				connection.close()
				connection = None
				break

			if dump:
				# save raw data away
				dump_fd.buffer.write(v)
				dump_fd.flush()

			if not p.do_hunk(rtr_session, v):
				break

def doit(args=None):
	"""RTR client"""

	debug = 0
	dump = False
	host = None
	port = None
	serial = None
	session_id = None
	timeout = 300 # five minutes for some random reason

	usage = (
					'usage: rtr_client '
					+ '[-H|--help] '
					+ '[-V|--version] '
					+ '[-v|--verbose] '
					+ '[-h HOSTNAME|--host=HOSTNAME] '
					+ '[-p PORTNUMBER|--port=PORTNUMBER] '
					+ '[-s SERIALNUMBER|--serial=SERIALNUMBER] '
					+ '[-S SESSIONID|--session=SESSIONID] '
					+ '[-t SECONDS|--timeout=SECONDS] '
					+ '[-d|--dump] '
		)

	try:
		opts, args = getopt.getopt(args, 'HVvh:p:s:S:t:d', [
						'help',
						'version',
						'verbose',
						'host=', 'port=',
						'serial=',
						'session=',
						'timeout=',
						'debug'
						])
	except getopt.GetoptError:
		sys.exit(usage)

	for opt, arg in opts:
		if opt in ('-H', '--help'):
			sys.exit(usage)
		if opt in ('-V', '--version'):
			sys.exit('%s: version: %s' % (sys.argv[0], __version__))
		elif opt in ('-v', '--verbose'):
			debug += 1
		elif opt in ('-h', '--host'):
			host = arg
		elif opt in ('-p', '--port'):
			port = int(arg)
		elif opt in ('-s', '--serial'):
			serial = int(arg)
		elif opt in ('-S', '--session'):
			session_id = int(arg)
		elif opt in ('-t', '--timeout'):
			timeout = int(arg)
		elif opt in ('-d', '--dump'):
			dump = True

	rtr_client(host=host, port=port, serial=serial, session_id=session_id, timeout=timeout, dump=dump, debug=debug)
	sys.exit(0)

def main(args=None):
	"""RTR client"""

	if args is None:
		args = sys.argv[1:]
	doit(args)

if __name__ == '__main__':
	main()

